### ISSUE TITLE
OpenURI: Make backend responsible for default applications

### ISSUE THREAD

**Issue by grulja** [flatpak/xdg-desktop-portal#472]
Right now you cannot change or affect what application will be associated with particular mimetype, it only works that way that when you use an app three times for a given mimetype, it will be automatically associated with it, which might not result into expected behavior. We have a nice configuration in KDE where we can configure applications which should be used based on mimetype, but this cannot be used as of now with xdg-desktop-portal. We should allow users to keep using existing system.

My idea is:
1) Backend implements a function, let's say something like "GetDefaultApplication(string mimetype)"
2) When xdg-desktop-portal gets a request to open a file, it can check whether the backend implements this function (based on version) and use default app provided by the backend
3) To make this secure, we will ask the user if he really wants to use this app for a given mimetype, this choice will be stored in the permission store so we don't ask again. If the backend returns a different application to the one we have saved before, then a new dialog would popup and asking again, to make sure we don't do something what wasn't approved by the user.
4) If the backend doesn't support this or the user rejects the default application provided by the backend, we will fallback to current behavior

Does this sound like a possible solution?

**Comment by hadess** [comment]
That sounds like a bandaid to me.

What explicit use cases are we trying to cover, and how are those handled on other platforms (such as iOS and Android)?

For most of the common "x-scheme-handle" mime-types, iOS and Android don't ask the user which handler they want to use when opening a link, but it uses the default browser (which might or might not be something the user can change, on iOS for example, that'd always be Safari). That'd also be 

The protection against the application opening thousands of URLs is that only one URL can be opened at a time, and the application can't request more URLs to be open if the application isn't active. So a rogue application can open one URL at most, and the user has the opportunity to kill/disable that app quickly.

For other data types, the OS usually only asks if there's more than one option.

So, what do the current use cases look like?

**Comment by hadess** [comment]
This is the upstream bug:
https://bugs.kde.org/show_bug.cgi?id=420050

The portal probably shouldn't do the "3 times and we stop asking" for browser URLs, and just always use the default browser for the platform. We'd need a way to stop rogue apps from spam-opening the browser though.

**Comment by grulja** [comment]
Just to make it clear, this is about OpenURI portal in general, not only OpenURI call as this also applies to OpenFile call as well.

**Comment by hadess** [comment]
> Just to make it clear, this is about OpenURI portal in general, not only OpenURI call as this also applies to OpenFile call as well.

Well understood, but we only have a use case for the OpenURI call, and none for the OpenFile one.

**Comment by grulja** [comment]
Use case can be:
You are used to open pictures with some editor, let's say GIMP and the portal will remember it. Then you switch to a different one, but the portal will keep using GIMP as it was associated already before, while for non-sandboxed apps you can easily set different default application.

**Comment by Erick555** [comment]
> The portal probably shouldn't do the "3 times and we stop asking" for browser URLs, and just always use the default browser for the platform.

What if I want to open link in non-default browser? Or doc in non-default pdf reader?

What I like in flatpak is that it allows to choose explicitly what app I want to launch instead of always launching the current default one which don't have to be my choice for particular file.

**Comment by hadess** [comment]
> What if I want to open link in non-default browser? Or doc in non-default pdf reader?

There's a "ask" option in the portal which the launching application can use.

**Comment by hadess** [comment]
>  You are used to open pictures with some editor, let's say GIMP and the portal will remember it. Then you switch to a different one, but the portal will keep using GIMP as it was associated already before, while for non-sandboxed apps you can easily set different default application.

I don't think the portal should remember it...

**Comment by hadess** [comment]
https://github.com/flatpak/xdg-desktop-portal/pull/473 solves the problem for internet URIs.

Which leaves the case of other file types. I think that it would be best if defaults were not remembered after some usage, but instead that we modify applications to either not ask, and open in the default application, or always ask. But even if that might not be easily feasible, we really need to get rid of that "click 3 times and it's your default" behaviour.

**Comment by grulja** [comment]
> > What if I want to open link in non-default browser? Or doc in non-default pdf reader?
> 
> There's a "ask" option in the portal which the launching application can use.

This option should be used only in case like file managers, where you have "Open With" option, but it shouldn't be the default when opening a file. Instead we should be opening what user expects to open and that is what he has configured in his system, like it does now for non-sandboxed apps.

**Comment by hadess** [comment]
> This option should be used only in case like file managers, where you have "Open With" option, but it shouldn't be the default when opening a file. Instead we should be opening what user expects to open and that is what he has configured in his system, like it does now for non-sandboxed apps.

I don't see why that should be restricted to file managers, or why applications can't use both. I wasn't saying that applications should always turn that on. But it's there if it needs to be used.

**Comment by grulja** [comment]
I didn't say it should be restricted to file managers, but many people will not like to keep picking application all the time when opening files, that's why I think it makes most sense in use case where you click on a file and choose "Open With" so you explicitly mention you want to pick an app. 

I agree that the portal should not remember the choice, but there should be a solution to open some default application and I don't see reason why we shouldn't use already existing system we have for example in KDE, it's just better integration for the sandboxed applications and in the solution I proposed at the beginning, I don't see any security issue. It's also opt-in, which means it would be us (KDE) who gets to decide if we want to use it this way or not, while if you don't consider this as a suitable solution for GNOME, you can keep the frontend to handle it as it does now.

**Comment by Erick555** [comment]
> It's also opt-in, which means it would be us (KDE) who gets to decide if we want to use it this way or not

I'll prefer users decide about it not developers.

**Comment by grulja** [comment]
> > It's also opt-in, which means it would be us (KDE) who gets to decide if we want to use it this way or not
> 
> I'll prefer users decide about it not developers.

You will decide whether to ask and choose app or not, but in case you don't want to be asked, then there needs to be way to pick a default application and that's what I'm trying to propose. Have backend implementation to handle this and not use the logic where when you use certain app three times, then it's considered as default.

**Comment by Pointedstick** [comment]
> I didn't say it should be restricted to file managers, but many people will not like to keep picking application all the time when opening files, that's why I think it makes most sense in use case where you click on a file and choose "Open With" so you explicitly mention you want to pick an app.

Exactly. It's really annoying to get asked the same question over and over again. It makes the system feel dumb; you expect it to remember your choice. But if it does this, then it needs a way for you to temporarily override that choice when needed.

We already have that for non-Flatpak apps and everyone likes it; the request here is to allow Flatpak to integrate with that so that the experience is seamless.

**Comment by nicolasfella** [comment]
Somewhat serious suggestion: xdg-desktop-portal calls and forwards the URI to xdg-open. That already takes into account platform differences and allows platforms to customize the behavior. it also would be perfectly consistent with non-Portal applications

**Comment by grulja** [comment]
Firefox bug for this issue: https://bugzilla.mozilla.org/show_bug.cgi?id=1621913

**Comment by matthiasclasen** [comment]
I don't really see how the fact that firefox can't make itself the default browser is related to the fact that the portal backend can't set the default browser.

And wrt to respecting system configuration - the portal frontend uses g_app_info_get_default_for_type, which relies on shared-mime configuration for defaults, so everything should work, as far as I can see.
